### ISSUE TITLE
Published binaries have main.version set via ldflags

### ISSUE THREAD

**Issue by jpalermo** [nats-io/nats-server#4598]
### What motivated this proposal?

The Syft scanner uses main.version to determine the version of a golang binary. This is not currently set on the published binaries, so it falls back to a composite version based on other data, i.e. `v0.0.0-20230906202347-528ee142cbe1`

### What is the proposed change?

Not sure where the published binaries are built, but they could incorporate the work that already happens in the [nightly docker file](https://github.com/nats-io/nats-server/blob/c6b26ab5d05d31c0bedf996ac1aedd7b2f590b2b/docker/Dockerfile.nightly#L10)

### Who benefits from this change?

Anybody generating SBOMs

**Comment by wallyqs** [comment]
Thanks for raising the issue, there is better metadata info that can be picked up via `go version -m nats-server` so thinking that ought to be picked up by syft instead but we can make the suggested change.

**Comment by jpalermo** [comment]
I don't currently see the version in that metadata.
```
go version -m ~/Downloads/nats-server-v2.10.1-linux-386/nats-server                                                                                                                                                                                                                                                   <master|>
/Users/jpalermo/Downloads/nats-server-v2.10.1-linux-386/nats-server: go1.21.1
	path	github.com/nats-io/nats-server/v2
	mod	github.com/nats-io/nats-server/v2	(devel)
	dep	github.com/klauspost/compress	v1.17.0	h1:Rnbp4K9EjcDuVuHtd0dgA4qNuv9yKDYKK1ulpJwgrqM=
	dep	github.com/minio/highwayhash	v1.0.2	h1:Aak5U0nElisjDCfPSG79Tgzkn2gl66NxOMspRrKnA/g=
	dep	github.com/nats-io/jwt/v2	v2.5.2	h1:DhGH+nKt+wIkDxM6qnVSKjokq5t59AZV5HRcFW0zJwU=
	dep	github.com/nats-io/nkeys	v0.4.5	h1:Zdz2BUlFm4fJlierwvGK+yl20IAKUm7eV6AAZXEhkPk=
	dep	github.com/nats-io/nuid	v1.0.1	h1:5iA8DT8V7q8WK2EScv2padNa/rTESc1KdnPw4TC2paw=
	dep	go.uber.org/automaxprocs	v1.5.3	h1:kWazyxZUrS3Gs4qUpbwo5kEIMGe/DAvi5Z4tl2NW4j8=
	dep	golang.org/x/crypto	v0.13.0	h1:mvySKfSWJ+UKUii46M40LOvyWfN0s2U+46/jDd0e6Ck=
	dep	golang.org/x/time	v0.3.0	h1:rg5rLMjNzMS1RkNLzCG38eapWhnYLFYXDXj2gOlr8j4=
	build	-buildmode=exe
	build	-compiler=gc
	build	-trimpath=true
	build	DefaultGODEBUG=panicnil=1
	build	CGO_ENABLED=0
	build	GOARCH=386
	build	GOOS=linux
	build	GO386=sse2
	build	vcs=git
	build	vcs.revision=d3ef745f253d1851b5c21e8502acb5d6a8e6eee7
	build	vcs.time=2023-09-20T17:20:51Z
	build	vcs.modified=false
```
but if `main.version` where set I think it would appear there.

**Comment by wallyqs** [comment]
I see, ok so setting `main.version` here would help knowing both the tag and the revision number

**Comment by jpalermo** [comment]
Yeah, syft does fall back to using the time and revision to make a pseudo-version, but it would be nice if the actual version/tag were in there too.

**Comment by wallyqs** [comment]
syft way of getting the version does not seem to work when using the `--trimpath` build option somehow, potentially a bug in Go unless I'm missing something...

**Comment by selzoc** [comment]
Huh, that's weird.  The [docs](https://pkg.go.dev/cmd/go#hdr-Compile_packages_and_dependencies) for `go build` mention nothing about that:
```
-trimpath
	remove all file system paths from the resulting executable.
	Instead of absolute file system paths, the recorded file names
	will begin either a module path@version (when using modules),
	or a plain import path (when using the standard library, or GOPATH).
```

Maybe they're assuming that all `-ldflags` contain paths?

For reference, here is the [syft code](https://github.com/anchore/syft/blob/38d5ef2c84e9bb8245c657daf2a3f97facde0a8f/syft/pkg/cataloger/golang/parse_go_binary.go#L99-L112) that tries to use the flags matching patterns, but then falls back to the `vcs.*` information.

**Comment by wallyqs** [comment]
likely syft should be using `module.Version` instead of `module.PseudoVersion` which is the generated one, but another catch is that `module.Version`  might end up still being `devel` when using `go build` (we would need to push the tag first then call go install explicitly with that tag to populate it).

**Comment by wallyqs** [comment]
So in the meantime, a workaround to avoid the pseudo version is to use `go install` and build the binary you need:

```
go install github.com/nats-io/nats-server/v2@v2.10.1
```

That would populate the version as needed for syft:

```
syft $GOPATH/bin/nats-server
 ✔ Cataloged packages              [10 packages]  
NAME                               VERSION  TYPE      
github.com/klauspost/compress      v1.17.0  go-module  
github.com/minio/highwayhash       v1.0.2   go-module  
github.com/nats-io/jwt/v2          v2.5.2   go-module  
github.com/nats-io/nats-server/v2  v2.10.1  go-module  
github.com/nats-io/nkeys           v0.4.5   go-module  
github.com/nats-io/nuid            v1.0.1   go-module  
go.uber.org/automaxprocs           v1.5.3   go-module  
golang.org/x/crypto                v0.13.0  go-module  
golang.org/x/sys                   v0.12.0  go-module  
golang.org/x/time                  v0.3.0   go-module
```

Ideally `--trimpath` would have worked but we could look into changing the way we populate this info by pushing a tag and then using `go install` instead of `go build` when building the binaries to avoid the module labeled as in using a `devel` version.

**Comment by wallyqs** [comment]
Go commit about trimpath removing ldflags: https://github.com/golang/go/commit/d252fdd630a51db206194b3c41c3d036fdd6f48a

**Comment by wallyqs** [comment]
Fixed via #5650 with a workaround since syft now inspects more metadata in the binary.
